### ISSUE TITLE
REST API for submitting DAG

### ISSUE THREAD

**Issue by karol-brejna-i** [gearpump/gearpump#1977]
1. 
Couldn't find description of endpoint for submitting GP application (http://www.gearpump.io/releases/latest/dev-rest-api.html)

2. 
I've noticed that the arguments ('arguments' field in submit form) are passed as URL encoded value which causes problems for longer values
Example: 
http://demo.gearpump.io/api/v1.0/master/submitapp?args=param1%20param2%20param3%20fdsafads%20param5

Can we pass them as field in the form? (So we avoid request size limit problem)

**Comment by kkasravi** [comment]
@karol-brejna-i 

```
1. Couldn't find description of endpoint for submitting GP application (http://www.gearpump.io/releases/latest/dev-rest-api.html)
```
We haven't formally documented this REST endpoint. See #1849.

```
2. I've noticed that the arguments ('arguments' field in submit form) are passed as URL encoded value which causes problems for longer values
Example: 
http://demo.gearpump.io/api/v1.0/master/submitapp?args=param1%20param2%20param3%20fdsafads%20param5

Can we pass them as field in the form? (So we avoid request size limit problem)
```
My recommendation is we include them in the JSON payload so we avoid the application/x-www-form-urlencoded mime-type. @stanleyxu2005 may have a preference.

**Comment by stanleyxu2005** [comment]
I think changing it as a form field would be a better choice. +1

**Comment by kkasravi** [comment]
In MasterService.scala we need to use formFields instead of parameters see http://doc.akka.io/docs/akka/2.4.2/scala/http/routing-dsl/index.html
For submit.js we need to provide a data field I believe to get the args sent as a form-field - https://github.com/danialfarid/ng-file-upload

@stanleyxu2005 would you be able to make these changes? Otherwise I'll make the changes and submit PR. I believe @karol-brejna-i needs the changes sometime Monday.

**Comment by stanleyxu2005** [comment]
The `ng-file-upload` I'm using has some difficulty to understand form field together with upload files. What concrete use case are you facing that cannot pass args in the url?

**Comment by kkasravi** [comment]
Hi @stanleyxu2005 

It does look like the latest version will help based on documentation I see under the `Upload Service`

https://github.com/danialfarid/ng-file-upload

```
  /*
  Specify the file and optional data to be sent to the server.
  Each field including nested objects will be sent as a form data multipart.
  Samples: {pic: file, username: username}
    {files: files, otherInfo: {id: id, person: person,...}} multiple files (html5)
    {profiles: {[{pic: file1, username: username1}, {pic: file2, username: username2}]} nested array multiple files (html5)
    {file: file, info: Upload.json({id: id, name: name, ...})} send fields as json string
    {file: file, info: Upload.jsonBlob({id: id, name: name, ...})} send fields as json blob, 'application/json' content_type
    {picFile: Upload.rename(file, 'profile.jpg'), title: title} send file with picFile key and profile.jpg file name*/
  *data: {key: file, otherInfo: uploadInfo},
```

**Comment by karol-brejna-i** [comment]
@stanleyxu2005, the parameters we are trying to pass are quite a large sections of json string (services credentials - kafka brokers, zookeeper servers, hbase credentials) that need to be used by a DAG.
Sending these as form parameters looked more "efficient". 
Maybe you can suggest some alternative if it's going to be a problem to implement...

**Comment by stanleyxu2005** [comment]
@karol-brejna-i Is it possible to upload a config file for this? I thought json is uploaded a config file in general.

**Comment by kkasravi** [comment]
I don't think passing this in as a config file will work since that is processed by the gear app. This is a specific argument passed to a GP application similar to what would happen on the command line. Normally a gearpump application invoked from the command line contacts the master directly and bypasses the services app. But in this case we want to send this information via the services app.

**Comment by stanleyxu2005** [comment]
I've minimized the issue and reported in #2009

**Comment by stanleyxu2005** [comment]
@karol-brejna-i the request has been implemented with #2023
